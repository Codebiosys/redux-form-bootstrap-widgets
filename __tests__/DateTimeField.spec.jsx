/* eslint-disable import/first */

import React from 'react';
import { mount, shallow } from 'enzyme';
import moment from 'moment';
import { DateTimeField } from 'index';

const fieldProps = {
  input: {
    name: 'fieldName',
    onChange: jest.fn(),
    onBlur: jest.fn(),
    onFocus: jest.fn(),
  },
  meta: {
    form: 'formName',
    pristine: false,
  },
  label: 'CheckBox  Test',
};

const selectedFieldProps = { ...fieldProps };
selectedFieldProps.input = { ...fieldProps.input };
selectedFieldProps.input.value = moment('2017-01-01');

describe('The Date Time Field', () => {
  let inputWrapper;

  beforeEach(() => {
    inputWrapper = mount(<DateTimeField {...fieldProps} />);
  });

  it('renders', () => {
    expect(inputWrapper).toMatchSnapshot();
  });

  it('has a label when there is a label', () => {
    expect(inputWrapper.find('ControlLabel').text()).toEqual(fieldProps.label);
  });

  it('does not display the clear button when there is no value', () => {
    expect(inputWrapper.find('.form-control-feedback').hasClass('glyphicon-calendar')).toBe(true);
  });


  it('clears the value when the clear button is pressed', () => {
    inputWrapper = mount(<DateTimeField {...selectedFieldProps} />);
    inputWrapper.find('.form-control-feedback').simulate('click');
    expect(selectedFieldProps.input.onChange).toHaveBeenCalledWith(null);
  });

  it('calls custom validator when the date changes', () => {
    const customValidator = jest.fn(() => ({ validationState: null, errorMessage: null }));
    const validatorProps = { ...fieldProps, validator: customValidator };
    const inputWrapperValidated = shallow(<DateTimeField {...validatorProps} />);
    expect(customValidator).toHaveBeenCalledTimes(1);
    inputWrapperValidated.setProps({ label: 'new Label' });
    expect(customValidator).toHaveBeenCalledTimes(2); // On Prop Change
  });

  it('renders the help message with a break', () => {
    const customValidator = jest.fn(() => ({ validationState: null, errorMessage: 'There was an error' }));
    const customProps = { ...fieldProps, helpText: 'The help Text', validator: customValidator };
    inputWrapper = shallow(<DateTimeField {...customProps} />);
    expect(inputWrapper).toMatchSnapshot();
  });
});
