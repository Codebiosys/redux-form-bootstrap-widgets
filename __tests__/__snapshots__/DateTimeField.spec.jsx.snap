// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`The Date Time Field renders 1`] = `
<DateTimeField
  disabled={false}
  input={
    Object {
      "name": "fieldName",
      "onBlur": [Function],
      "onChange": [Function],
      "onFocus": [Function],
    }
  }
  label="CheckBox  Test"
  meta={
    Object {
      "form": "formName",
      "pristine": false,
    }
  }
  required={false}
  validator={[Function]}
>
  <FormGroup
    bsClass="form-group"
    controlId="fieldName"
  >
    <div
      className="form-group"
    >
      <Label
        label="CheckBox  Test"
        required={false}
      >
        <ControlLabel
          bsClass="control-label"
          srOnly={false}
        >
          <label
            className="control-label"
            htmlFor="fieldName"
          >
            CheckBox  Test
          </label>
        </ControlLabel>
      </Label>
      <Component
        className=""
        closeOnSelect={true}
        closeOnTab={true}
        dateFormat={true}
        defaultValue=""
        id="formName-fieldName"
        input={true}
        inputProps={Object {}}
        name="fieldName"
        onBlur={[Function]}
        onChange={[Function]}
        onFocus={[Function]}
        onViewModeChange={[Function]}
        renderInput={[Function]}
        strictParsing={true}
        timeConstraints={Object {}}
        timeFormat={true}
        utc={false}
        validator={[Function]}
      >
        <div
          className="rdt"
        >
          <div>
            <InputGroup
              bsClass="input-group"
            >
              <span
                className="input-group"
              >
                <FormControl
                  autoComplete="off"
                  bsClass="form-control"
                  className="form-control"
                  componentClass="input"
                  disabled={false}
                  onChange={[Function]}
                  onClick={[Function]}
                  onFocus={[Function]}
                  onKeyDown={[Function]}
                  type="text"
                >
                  <input
                    autoComplete="off"
                    className="form-control form-control"
                    disabled={false}
                    id="fieldName"
                    onChange={[Function]}
                    onClick={[Function]}
                    onFocus={[Function]}
                    onKeyDown={[Function]}
                    type="text"
                  />
                </FormControl>
                <FormControlFeedback
                  bsClass="form-control-feedback"
                  bsRole="feedback"
                  style={
                    Object {
                      "pointerEvents": "none",
                    }
                  }
                >
                  <Glyphicon
                    bsClass="glyphicon"
                    className="form-control-feedback"
                    glyph="calendar"
                    style={
                      Object {
                        "pointerEvents": "none",
                      }
                    }
                  >
                    <span
                      className="form-control-feedback glyphicon glyphicon-calendar"
                      style={
                        Object {
                          "pointerEvents": "none",
                        }
                      }
                    />
                  </Glyphicon>
                </FormControlFeedback>
              </span>
            </InputGroup>
          </div>
          <div
            className="rdtPicker"
          >
            <Component
              onClickOutside={[Function]}
              view="days"
              viewProps={
                Object {
                  "addTime": [Function],
                  "dateFormat": "MM/DD/YYYY",
                  "handleClickOutside": [Function],
                  "isValidDate": undefined,
                  "localMoment": [Function],
                  "renderDay": undefined,
                  "renderMonth": undefined,
                  "renderYear": undefined,
                  "selectedDate": undefined,
                  "setDate": [Function],
                  "setTime": [Function],
                  "showView": [Function],
                  "subtractTime": [Function],
                  "timeConstraints": Object {},
                  "timeFormat": "h:mm A",
                  "updateOn": "days",
                  "updateSelectedDate": [Function],
                  "value": undefined,
                  "viewDate": "2018-02-01T08:00:00.000Z",
                }
              }
            >
              <OnClickOutside(Component)
                addTime={[Function]}
                dateFormat="MM/DD/YYYY"
                eventTypes={
                  Array [
                    "mousedown",
                    "touchstart",
                  ]
                }
                excludeScrollbar={false}
                handleClickOutside={[Function]}
                localMoment={[Function]}
                outsideClickIgnoreClass="ignore-react-onclickoutside"
                preventDefault={false}
                setDate={[Function]}
                setTime={[Function]}
                showView={[Function]}
                stopPropagation={false}
                subtractTime={[Function]}
                timeConstraints={Object {}}
                timeFormat="h:mm A"
                updateOn="days"
                updateSelectedDate={[Function]}
                viewDate={"2018-02-01T08:00:00.000Z"}
              >
                <Component
                  addTime={[Function]}
                  dateFormat="MM/DD/YYYY"
                  disableOnClickOutside={[Function]}
                  enableOnClickOutside={[Function]}
                  eventTypes={
                    Array [
                      "mousedown",
                      "touchstart",
                    ]
                  }
                  handleClickOutside={[Function]}
                  localMoment={[Function]}
                  outsideClickIgnoreClass="ignore-react-onclickoutside"
                  preventDefault={false}
                  setDate={[Function]}
                  setTime={[Function]}
                  showView={[Function]}
                  stopPropagation={false}
                  subtractTime={[Function]}
                  timeConstraints={Object {}}
                  timeFormat="h:mm A"
                  updateOn="days"
                  updateSelectedDate={[Function]}
                  viewDate={"2018-02-01T08:00:00.000Z"}
                >
                  <div
                    className="rdtDays"
                  >
                    <table>
                      <thead>
                        <tr>
                          <th
                            className="rdtPrev"
                            onClick={[Function]}
                          >
                            <span>
                              ‹
                            </span>
                          </th>
                          <th
                            className="rdtSwitch"
                            colSpan={5}
                            data-value={1}
                            onClick={[Function]}
                          >
                            February 2018
                          </th>
                          <th
                            className="rdtNext"
                            onClick={[Function]}
                          >
                            <span>
                              ›
                            </span>
                          </th>
                        </tr>
                        <tr>
                          <th
                            className="dow"
                          >
                            Su
                          </th>
                          <th
                            className="dow"
                          >
                            Mo
                          </th>
                          <th
                            className="dow"
                          >
                            Tu
                          </th>
                          <th
                            className="dow"
                          >
                            We
                          </th>
                          <th
                            className="dow"
                          >
                            Th
                          </th>
                          <th
                            className="dow"
                          >
                            Fr
                          </th>
                          <th
                            className="dow"
                          >
                            Sa
                          </th>
                        </tr>
                      </thead>
                      <tbody>
                        <tr>
                          <td
                            className="rdtDay rdtOld"
                            data-value={28}
                            onClick={[Function]}
                          >
                            28
                          </td>
                          <td
                            className="rdtDay rdtOld"
                            data-value={29}
                            onClick={[Function]}
                          >
                            29
                          </td>
                          <td
                            className="rdtDay rdtOld"
                            data-value={30}
                            onClick={[Function]}
                          >
                            30
                          </td>
                          <td
                            className="rdtDay rdtOld"
                            data-value={31}
                            onClick={[Function]}
                          >
                            31
                          </td>
                          <td
                            className="rdtDay"
                            data-value={1}
                            onClick={[Function]}
                          >
                            1
                          </td>
                          <td
                            className="rdtDay"
                            data-value={2}
                            onClick={[Function]}
                          >
                            2
                          </td>
                          <td
                            className="rdtDay"
                            data-value={3}
                            onClick={[Function]}
                          >
                            3
                          </td>
                        </tr>
                        <tr>
                          <td
                            className="rdtDay"
                            data-value={4}
                            onClick={[Function]}
                          >
                            4
                          </td>
                          <td
                            className="rdtDay rdtToday"
                            data-value={5}
                            onClick={[Function]}
                          >
                            5
                          </td>
                          <td
                            className="rdtDay"
                            data-value={6}
                            onClick={[Function]}
                          >
                            6
                          </td>
                          <td
                            className="rdtDay"
                            data-value={7}
                            onClick={[Function]}
                          >
                            7
                          </td>
                          <td
                            className="rdtDay"
                            data-value={8}
                            onClick={[Function]}
                          >
                            8
                          </td>
                          <td
                            className="rdtDay"
                            data-value={9}
                            onClick={[Function]}
                          >
                            9
                          </td>
                          <td
                            className="rdtDay"
                            data-value={10}
                            onClick={[Function]}
                          >
                            10
                          </td>
                        </tr>
                        <tr>
                          <td
                            className="rdtDay"
                            data-value={11}
                            onClick={[Function]}
                          >
                            11
                          </td>
                          <td
                            className="rdtDay"
                            data-value={12}
                            onClick={[Function]}
                          >
                            12
                          </td>
                          <td
                            className="rdtDay"
                            data-value={13}
                            onClick={[Function]}
                          >
                            13
                          </td>
                          <td
                            className="rdtDay"
                            data-value={14}
                            onClick={[Function]}
                          >
                            14
                          </td>
                          <td
                            className="rdtDay"
                            data-value={15}
                            onClick={[Function]}
                          >
                            15
                          </td>
                          <td
                            className="rdtDay"
                            data-value={16}
                            onClick={[Function]}
                          >
                            16
                          </td>
                          <td
                            className="rdtDay"
                            data-value={17}
                            onClick={[Function]}
                          >
                            17
                          </td>
                        </tr>
                        <tr>
                          <td
                            className="rdtDay"
                            data-value={18}
                            onClick={[Function]}
                          >
                            18
                          </td>
                          <td
                            className="rdtDay"
                            data-value={19}
                            onClick={[Function]}
                          >
                            19
                          </td>
                          <td
                            className="rdtDay"
                            data-value={20}
                            onClick={[Function]}
                          >
                            20
                          </td>
                          <td
                            className="rdtDay"
                            data-value={21}
                            onClick={[Function]}
                          >
                            21
                          </td>
                          <td
                            className="rdtDay"
                            data-value={22}
                            onClick={[Function]}
                          >
                            22
                          </td>
                          <td
                            className="rdtDay"
                            data-value={23}
                            onClick={[Function]}
                          >
                            23
                          </td>
                          <td
                            className="rdtDay"
                            data-value={24}
                            onClick={[Function]}
                          >
                            24
                          </td>
                        </tr>
                        <tr>
                          <td
                            className="rdtDay"
                            data-value={25}
                            onClick={[Function]}
                          >
                            25
                          </td>
                          <td
                            className="rdtDay"
                            data-value={26}
                            onClick={[Function]}
                          >
                            26
                          </td>
                          <td
                            className="rdtDay"
                            data-value={27}
                            onClick={[Function]}
                          >
                            27
                          </td>
                          <td
                            className="rdtDay"
                            data-value={28}
                            onClick={[Function]}
                          >
                            28
                          </td>
                          <td
                            className="rdtDay rdtNew"
                            data-value={1}
                            onClick={[Function]}
                          >
                            1
                          </td>
                          <td
                            className="rdtDay rdtNew"
                            data-value={2}
                            onClick={[Function]}
                          >
                            2
                          </td>
                          <td
                            className="rdtDay rdtNew"
                            data-value={3}
                            onClick={[Function]}
                          >
                            3
                          </td>
                        </tr>
                        <tr>
                          <td
                            className="rdtDay rdtNew"
                            data-value={4}
                            onClick={[Function]}
                          >
                            4
                          </td>
                          <td
                            className="rdtDay rdtNew"
                            data-value={5}
                            onClick={[Function]}
                          >
                            5
                          </td>
                          <td
                            className="rdtDay rdtNew"
                            data-value={6}
                            onClick={[Function]}
                          >
                            6
                          </td>
                          <td
                            className="rdtDay rdtNew"
                            data-value={7}
                            onClick={[Function]}
                          >
                            7
                          </td>
                          <td
                            className="rdtDay rdtNew"
                            data-value={8}
                            onClick={[Function]}
                          >
                            8
                          </td>
                          <td
                            className="rdtDay rdtNew"
                            data-value={9}
                            onClick={[Function]}
                          >
                            9
                          </td>
                          <td
                            className="rdtDay rdtNew"
                            data-value={10}
                            onClick={[Function]}
                          >
                            10
                          </td>
                        </tr>
                      </tbody>
                      <tfoot>
                        <tr>
                          <td
                            className="rdtTimeToggle"
                            colSpan={7}
                            onClick={[Function]}
                          >
                            12:00 AM
                          </td>
                        </tr>
                      </tfoot>
                    </table>
                  </div>
                </Component>
              </OnClickOutside(Component)>
            </Component>
          </div>
        </div>
      </Component>
      <HelpBlock
        bsClass="help-block"
        style={
          Object {
            "minHeight": "3ex",
          }
        }
      >
        <span
          className="help-block"
          style={
            Object {
              "minHeight": "3ex",
            }
          }
        />
      </HelpBlock>
    </div>
  </FormGroup>
</DateTimeField>
`;

exports[`The Date Time Field renders the help message with a break 1`] = `
<FormGroup
  bsClass="form-group"
  controlId="fieldName"
  validationState={null}
>
  <Label
    label="CheckBox  Test"
    required={false}
  />
  <Component
    className=""
    closeOnSelect={true}
    closeOnTab={true}
    dateFormat={true}
    defaultValue=""
    id="formName-fieldName"
    input={true}
    inputProps={Object {}}
    name="fieldName"
    onBlur={[Function]}
    onChange={[Function]}
    onFocus={[Function]}
    onViewModeChange={[Function]}
    renderInput={[Function]}
    strictParsing={true}
    timeConstraints={Object {}}
    timeFormat={true}
    utc={false}
    validator={[Function]}
  />
  <HelpBlock
    bsClass="help-block"
    style={
      Object {
        "minHeight": "6ex",
      }
    }
  >
    There was an error
    <br />
    The help Text
  </HelpBlock>
</FormGroup>
`;
